class.jetpack.php

```diff
@@ -738,11 +732,65 @@ function( $callable ) {
 		add_filter( 'jetpack_token_request_body', array( __CLASS__, 'filter_token_request_body' ) );
 	}
 
+	/**
+	 * Runs on plugins_loaded but with a higher priority to make sure we're initializing
+	 * before other hooks.
+	 *
+	 * @action plugins_loaded
+	 * @priority 5
+	 */
+	public function early_initalization() {
+		/*
+		 * Enable enhanced handling of previewing sites in Calypso
+		 */
+		if ( self::is_active() ) {
+			require_once JETPACK__PLUGIN_DIR . '_inc/lib/class.jetpack-iframe-embed.php';
+			add_action( 'init', array( 'Jetpack_Iframe_Embed', 'init' ), 9, 0 );
+			require_once JETPACK__PLUGIN_DIR . '_inc/lib/class.jetpack-keyring-service-helper.php';
+			add_action( 'init', array( 'Jetpack_Keyring_Service_Helper', 'init' ), 9, 0 );
+		}
+
+		Sync_Main::init();
```

Comment:
By doing that, we will no longer be able to sync anything that happens between the Jetpack plugin file load and the `plugins_loaded` hook. While in theory there shouldn't be much happening there, in practice I'm not sure we're not taking advantage of the previous way it was written somehow.

Reply:
The only potential change this brings is that the two events that are added in `late_initialization` now will only catch events after `plugins_loaded`. Namely these are `updating_jetpack_version` and `jetpack_user_authorized`. Off the top of your head do you know if these two can happen before `plugins_loaded`?
Everything else is happening exactly as it was before - on `plugins_loaded` with the same priority.

Reply:
Not sure about any of those without checking specifically. But on the top of my head I can recall of multiple instances where classes from packages are requested before plugins have been loaded - remember those ealry usage notices generated by the autoloader?

Reply:
> remember those ealry usage notices generated by the autoloader?

I vote for leading by the example and not do anything before `plugins_loaded`, if a site doesn't sync something because it happens before `plugins_loaded` they should fix whatever does that change.